src/framework/MOM_intrinsic_functions.F90

```diff
@@ -25,4 +28,127 @@ function invcosh(x)
+      ! at the end, and it is therefore more computationally efficient.
+
+      ! Because successive estimates of the numerator and denominator tend to be the cube of their
+      ! predecessors, the numerator and denominator need to be rescaled by division when they get
+      ! too large or small to avoid overflow or underflow in the convergence test below.  Were
+      ! this being done for 32 bit reals, the values to compare with would be about 1.0e9 and 1.0e-9.
+      if ((den > 1.0e75) .or. (den < 1.0e-75)) then
+        num = num / den ; den = 1.0
+      endif
+
+      num_prev = num ; den_prev = den
+      num = (2.0 * num_prev**3 + asx * den_prev**3)
+      den = 3.0 * (den_prev * num_prev**2)
+
+      !   If successive estimates of root are identical, this is an exact or converged solution.
+      if ((num * den_prev == num_prev * den) .or. (itt == 9)) then
+        root_asx = num / den
+        exit
+      endif
+
+      !   If the estimates are increasing, this also indicates convergence, but for a more subtle
+      ! reason.  Because Newton's method converges monotonically from above (at least for infinite
+      ! precision math), the only reason why this estimate could increase is if the iterations
+      ! have converged to a roundoff-level limit cycle around an irrational or otherwise
+      ! unrepresentable solution, with values only changing in the last bit or two.  If so, we
+      ! should stop iterating and accept the one of the current or previous solutions, both of
+      ! which will be within numerical roundoff of the true solution.
+      if (num * den_prev > num_prev * den) then
+        root_asx = num / den
+        ! Pick the more accurate of the last two iterations.
+        ! Given that both of the two previous iterations are within roundoff of the true
+        ! solution, this next step might be overkill.
+        if ( abs(den_prev**3*root_asx**3 - den_prev**3*asx) > abs(num_prev**3 - den_prev**3*asx) ) then
+          ! The previous iteration was slightly more accurate, so use that for root_asx.
+          root_asx = num_prev / den_prev
+        endif
+        exit
+      endif
+    enddo
+
+    root = sign(scale(root_asx, ex_3), x)
+  endif
+
+end function cuberoot
+
+!> Returns true if any unit test of intrinsic_functions fails, or false if they all pass.
+logical function intrinsic_functions_unit_tests(verbose)
+  logical, intent(in) :: verbose !< If true, write results to stdout
+
+  ! Local variables
+  real :: testval  ! A test value for self-consistency testing [nondim]
+  logical :: fail, v
+
+  fail = .false.
+  v = verbose
+  write(stdout,*) '==== MOM_intrinsic_functions: intrinsic_functions_unit_tests ==='
+
+  fail = fail .or. Test_cuberoot(v, 1.2345678901234e9)
+  fail = fail .or. Test_cuberoot(v, -9.8765432109876e-21)
+  fail = fail .or. Test_cuberoot(v, 64.0)
+  fail = fail .or. Test_cuberoot(v, -0.5000000000001)
+  fail = fail .or. Test_cuberoot(v, 0.0)
```

Comment:
I would add the following tests, since they are important boundary values.
```
fail = fail .or. Test_cuberoot(v, 0.125)
fail = fail .or. Test_cuberoot(v, 1.0)
```

Reply:
Done.